packages/ember-glimmer/lib/renderer.js

```diff
@@ -197,83 +123,131 @@ class Renderer {
       rootOutletState: UNDEFINED_REFERENCE,
       isTopLevel: true
     });
-
-    let result, shouldReflush, afterRender, didChangeAfterRender;
-
-    let callback = () => {
-      result = view.template.asEntryPoint().render(self, env, { appendTo: target, dynamicScope });
-    };
-
-    let rerender = () => result.rerender({ alwaysRevalidate: shouldReflush });
-
-    do {
-      env.begin();
-      shouldReflush = runInTransaction(callback);
-      afterRender = CURRENT_TAG.value();
-      env.commit();
-      didChangeAfterRender = !CURRENT_TAG.validate(afterRender);
-      callback = rerender;
-    } while (shouldReflush || didChangeAfterRender);
-
-    this._scheduler.registerView(view);
-
-    // FIXME: Store this somewhere else
-    view['_renderResult'] = result;
-
-    // FIXME: This should happen inside `env.commit()`
-    view._transitionTo('inDOM');
+    this._renderRoot(view, view.template, self, target, dynamicScope);
   }
 
   rerender(view) {
-    let { _env: env } = this;
+    this._scheduleRevalidate();
+  }
+
+  componentInitAttrs() {
+    // TODO: Remove me
+  }
 
-    let renderResult = view['_renderResult'] || this._root['_renderResult'];
-    let shouldReflush = false;
-    let afterRender, didChangeAfterRender;
+  ensureViewNotRendering() {
+    // TODO: Implement this
+    // throw new Error('Something you did caused a view to re-render after it rendered but before it was inserted into the DOM.');
+  }
 
-    let callback = () => renderResult.rerender({ alwaysRevalidate: shouldReflush });
+  register(view) {
+    assert('Attempted to register a view with an id already in use: ' + view.elementId, !this._viewRegistry[this.elementId]);
+    this._viewRegistry[view.elementId] = view;
+  }
 
-    do {
-      env.begin();
-      shouldReflush = runInTransaction(callback);
-      afterRender = CURRENT_TAG.value();
-      env.commit();
-      didChangeAfterRender = !CURRENT_TAG.validate(afterRender);
-    } while (shouldReflush || didChangeAfterRender);
+  unregister(view) {
+    delete this._viewRegistry[this.elementId];
   }
 
   remove(view) {
-    this._scheduler.deregisterView(view);
     view.trigger('willDestroyElement');
     view._transitionTo('destroying');
 
-    let { _renderResult } = view;
-
-    if (_renderResult) {
-      _renderResult.destroy();
+    if (this._root === view) {
+      this._clearRoot();
     }
 
     if (!view.isDestroying) {
       view.destroy();
     }
   }
 
-  componentInitAttrs() {
-    // TODO: Remove me
+  destroy() {
+    if (this._destroyed) {
+      return;
+    }
+    this._destroyed = true;
+    this._clearRoot();
   }
 
-  ensureViewNotRendering() {
-    // TODO: Implement this
-    // throw new Error('Something you did caused a view to re-render after it rendered but before it was inserted into the DOM.');
+  _renderRoot(root, template, self, target, dynamicScope) {
+    assert(!this._root, 'root is not in DOM');
```

Comment:
The arguments are flipped (also, I prefer "cannot append multiple root views" or something like that)